overlord/ifacestate/handlers.go

```diff
@@ -487,6 +487,13 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
 	if err != nil || conn == nil {
 		return err
 	}
+	defer func() {
+		if err != nil {
+			if err := m.repo.Disconnect(plugRef.Snap, plugRef.Name, slotRef.Snap, slotRef.Name); err != nil {
```

Comment:
Just to be 100% sure: the err you define here is different from the one we'll return, right? It must be so, otherwise the test wouldn't pass, but I'm a bit confused because usually when I do a `:=` on a variable which already exists the compiler complains that no new variables are being defined.

Reply:
+1, it's not very clear if we are not shadowing err in this function, would it make sense to use a different name for the err variable that we don't shadowed?

Reply:
@mardy just to (try to) answer your question: what you're mentioning is a redeclaration in the same scope. In this case, the variable is being redeclared in a new nested scope (shadowing) which creates a new variable. This new variable can't be accessed out of the scope, so outside the `if err != nil` block, err will have the same value as before

Reply:
@MiguelPires thanks, it's clear now! :-)

Reply:
The shadowing is fine here, it's a feature of go that it works https://golang.org/ref/spec#Defer_statements:
```
Deferred functions are executed after any result parameters are set by that return statement but before the function returns to its caller.
```
We use this idiom a lot already in similar situations, e.g. in https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1293

Reply:
yes, for that reason I'd consider renaming err to something else to make it very clear on what error do we actually want  to act with the deferred cleanup.

Reply:
@mvo, sorry, I didn't mean it's wrong (it's all correct), what I mean is it's not immediately obvious that `if err := m.setupSnapSecurity(task, slot.Snap, slotOpts, perfTimings...` below (which we are interested in defer) is actually setting the err that the defer is about when deciding to do Disconnect. Maybe renaming the main err to something like setupErr would highlight it. Does it make sense or am I confused?

Reply:
my impression in general is that although golang scoping rules with defer etc are a bit complicated, introducing non-standard names for error is always a bit confusing for the reader in itself